notebooks/predicting_queries/simple_intent_suggestion.ipynb

```diff
@@ -0,0 +1,570 @@
+    "service_terms= set(title_df['body_tokens'].explode('body_tokens'))\n",
+    "# Lose some stop words\n",
+    "intersection = service_terms.difference(set(['i','of','the','a','and','to','you','if','that','then','in','on']))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "title_df['intersect_tokens'] = title_df['body_tokens'].map(lambda x: [token for token in x if token in intersection])\n",
+    "title_df['body_token_count'] = title_df['intersect_tokens'].map(lambda x: Counter(x))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# bag of words for mainstream content\n",
+    "service_bow = pd.DataFrame.from_records(title_df['body_token_count'],index=title_df['name'])\n",
+    "service_bow = service_bow.replace(np.nan,0)\n",
+    "#add one smoothing\n",
+    "service_bow = service_bow +1\n",
+    "service_dis = service_bow.divide(service_bow.sum(1),0)\n",
+    "service_dis.shape"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "query_taxons = query_taxons[['taxon','tokens']]\n",
+    "query_taxons['valid_tokens'] = query_taxons['tokens'].map(lambda x: [token for token in x if token in intersection])\n",
+    "query_taxons['valid_token_count'] = query_taxons['valid_tokens'].map(lambda x:Counter(x))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Bag of words for taxon queries\n",
+    "query_bow = pd.DataFrame.from_records(query_taxons['valid_token_count'],index=query_taxons['taxon'],columns=service_dis.columns)\n",
+    "# Don't smooth unless you have lots of data\n",
+    "query_bow = query_bow.replace(np.nan,0)\n",
+    "\n",
+    "query_dis = query_bow.divide(query_bow.sum(1),0)\n",
+    "query_dis.shape\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "assert(query_dis.shape[1]==service_dis.shape[1]==len(intersection))\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# Find min KL divergence content to taxon query terms\n",
+    "# https://en.wikipedia.org/wiki/Information_projection\n",
+    "# Add one smoothing, which is p, which is q, all make quite a big difference in performance\n",
+    "# p is our normalised bag of words from each taxon query\n",
+    "# q is every bit of mainstream content normalised bag of words\n",
+    "# e.g try 'Blue badges', 'Afghanistan' etc\n",
+    "TAXON = \"Stopping or selling your business\"\n",
+    "idx = np.argsort(entropy(np.broadcast_to(query_dis.loc[TAXON].to_numpy(),(service_dis.shape[0],service_dis.shape[1])),service_dis,axis=1))[0:15]\n",
```

Comment:
This is quite difficult to understand, especially without named arguments. Possibly better to split out into separate lines so it is easier to follow the logic.
```suggestion
    "# get entropy of a distribution for given probability values\n"
    "ent = entropy(pk=np.broadcast_to(query_dis.loc[TAXON].to_numpy(),\n"
                            "qk=(service_dis.shape[0],service_dis.shape[1])),\n"
                            "base=service_dis,\n"
                            "axis=1)\n"
    "# get indices to sort array\n"
    "idx = np.argsort(a=ent)[0:15]\n"
```

Reply:
yep i'll tweak that!